rust/lance/src/index/scalar.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Utilities for integrating scalar indices with datasets
+//!
+
+use std::{future, sync::Arc};
+
+use arrow::compute::kernels;
+use arrow_array::RecordBatch;
+use async_trait::async_trait;
+use futures::{stream::BoxStream, StreamExt, TryStreamExt};
+use lance_index::scalar::{
+    btree::{train_btree_index, BTreeIndex, BtreeTrainingSource},
+    flat::FlatIndexTrainer,
+    lance_format::LanceIndexStore,
+    ScalarIndex,
+};
+use snafu::{location, Location};
+use tracing::instrument;
+
+use lance_core::{Error, Result};
+
+use crate::{
+    dataset::{chunker, scanner::ColumnOrdering},
+    Dataset,
+};
+
+use super::IndexParams;
+
+#[derive(Default)]
+pub struct ScalarIndexParams {}
+
+impl IndexParams for ScalarIndexParams {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+}
+
+struct TrainingRequest {
+    dataset: Arc<Dataset>,
+    column: String,
+}
+
+#[async_trait]
+impl BtreeTrainingSource for TrainingRequest {
+    async fn scan_ordered_chunks(
+        self: Box<Self>,
+        chunk_size: u32,
+    ) -> Result<BoxStream<'static, Result<RecordBatch>>> {
+        let mut scan = self.dataset.scan();
+        let scan = scan
+            .with_row_id()
+            .order_by(Some(vec![ColumnOrdering::asc_nulls_first(
+                self.column.clone(),
+            )]))?
+            .project(&[&self.column])?;
+
+        let schema = scan.schema()?;
+        let ordered_batches = scan.try_into_stream().await?;
+        let chunked = chunker::chunk_stream(ordered_batches.into(), chunk_size as usize);
+        Ok(chunked
+            .and_then(move |batches| {
+                future::ready(
+                    // chunk_stream is zero-copy and so it gives us pieces of batches.  However, the btree
+                    // index needs 1 batch-per-page and so we concatenate here.
+                    kernels::concat::concat_batches(&schema, batches.iter()).map_err(|e| e.into()),
+                )
+            })
+            .boxed())
+    }
+}
+
+/// Build a Vector Index
+#[instrument(skip(dataset))]
+pub async fn build_scalar_index(dataset: &Dataset, column: &str, uuid: &str) -> Result<()> {
```

Comment:
This is sad that we still have these functions in `lance` 😵‍💫

Reply:
I think some of the vector functionality can move out of `lance` but not all.  I think the functionality in this file has to be in `lance`:

 * Index creation
   * The manifest update will need to be in `lance`
   * Any scanning of the source data will need to be in `lance` (I've abstracted this away with `BtreeTrainingSource`)
   * Any reading or writing of index data will need to know what directory to read and write.  (I've abstracted this away from `IndexStore`)